x-pack/platform/packages/shared/kbn-inference-endpoint-ui-common/src/components/inference_service_form_fields.tsx

```diff
@@ -0,0 +1,357 @@
+
+interface InferenceServicesProps {
+  providers: InferenceProvider[];
+}
+
+export const InferenceServiceFormFields: React.FC<InferenceServicesProps> = ({ providers }) => {
+  const [isProviderPopoverOpen, setProviderPopoverOpen] = useState(false);
+  const [providerSchema, setProviderSchema] = useState<ConfigEntryView[]>([]);
+  const [taskTypeOptions, setTaskTypeOptions] = useState<TaskTypeOption[]>([]);
+  const [selectedTaskType, setSelectedTaskType] = useState<string>(DEFAULT_TASK_TYPE);
+
+  const { updateFieldValues, setFieldValue, validateFields, isSubmitting } = useFormContext();
+  const [requiredProviderFormFields, setRequiredProviderFormFields] = useState<ConfigEntryView[]>(
+    []
+  );
+  const [optionalProviderFormFields, setOptionalProviderFormFields] = useState<ConfigEntryView[]>(
+    []
+  );
+  const [{ config, secrets }] = useFormData<ConnectorFormSchema<Config, Secrets>>({
+    watch: [
+      'secrets.providerSecrets',
+      'config.taskType',
+      'config.inferenceId',
+      'config.provider',
+      'config.providerConfig',
+    ],
+  });
+
+  const handleProviderPopover = useCallback(() => {
+    setProviderPopoverOpen((isOpen) => !isOpen);
+  }, []);
+
+  const handleProviderClosePopover = useCallback(() => {
+    setProviderPopoverOpen(false);
+  }, []);
+
+  const handleProviderKeyboardOpen: EuiFieldTextProps['onKeyDown'] = useCallback((event: any) => {
+    if (event.key === keys.ENTER) {
+      setProviderPopoverOpen(true);
+    }
+  }, []);
+
+  const providerIcon = useMemo(
+    () =>
+      Object.keys(SERVICE_PROVIDERS).includes(config?.provider)
+        ? SERVICE_PROVIDERS[config?.provider as ServiceProviderKeys].icon
+        : undefined,
+    [config?.provider]
+  );
+
+  const providerName = useMemo(
+    () =>
+      Object.keys(SERVICE_PROVIDERS).includes(config?.provider)
+        ? SERVICE_PROVIDERS[config?.provider as ServiceProviderKeys].name
+        : config?.provider,
+    [config?.provider]
+  );
+
+  const onTaskTypeOptionsSelect = useCallback(
+    (taskType: string) => {
+      setSelectedTaskType(taskType);
+
+      const inferenceId = generateInferenceEndpointId({
+        ...config,
+        taskType,
+      });
+
+      updateFieldValues({
+        config: {
+          taskType,
+          inferenceId,
+        },
+      });
+    },
+    [config, updateFieldValues]
+  );
+
+  const onProviderChange = useCallback(
+    (provider?: string) => {
+      const newProvider = providers?.find((p) => p.service === provider);
+
+      setTaskTypeOptions(getTaskTypeOptions(newProvider?.task_types ?? []));
+      if (newProvider?.task_types && newProvider?.task_types.length > 0) {
+        onTaskTypeOptionsSelect(newProvider?.task_types[0]);
+      }
+
+      // Update providerSchema
+      const newProviderSchema = Object.keys(newProvider?.configurations ?? {}).map((k) => ({
+        key: k,
+        isValid: true,
+        ...newProvider?.configurations[k],
+      })) as ConfigEntryView[];
```

Comment:
Not only this but in general, can we do this transformations without `as`? what happens if you set return type `ConfigEntryView` to the arrow function in `map`? 

currently we override whatever it is and cast the type to `ConfigEntryView[]` and missing errors to be introduced with a change.